metadata/src/audio/item.rs

```diff
@@ -16,98 +18,290 @@ use librespot_core::{
+}
+
+#[derive(Debug, Clone)]
+pub enum UniqueFields {
+    Track {
+        artists: Vec<String>,
+        album: String,
+        album_artists: Vec<String>,
+        popularity: u8,
+        number: u32,
+        disc_number: u32,
+    },
+    Episode {
+        description: String,
+        publish_time: Date,
+        show_name: String,
+    },
 }
 
 impl AudioItem {
     pub async fn get_file(session: &Session, id: SpotifyId) -> AudioItemResult {
+        let image_url = session
+            .get_user_attribute("image-url")
+            .unwrap_or_else(|| String::from("https://i.scdn.co/image/{file_id}"));
+
         match id.item_type {
-            SpotifyItemType::Track => Track::get_audio_item(session, id).await,
-            SpotifyItemType::Episode => Episode::get_audio_item(session, id).await,
-            _ => Err(Error::unavailable(MetadataError::NonPlayable)),
-        }
-    }
-}
+            SpotifyItemType::Track => {
+                let track = Track::get(session, &id).await?;
 
-#[async_trait]
-pub trait InnerAudioItem {
-    async fn get_audio_item(session: &Session, id: SpotifyId) -> AudioItemResult;
-
-    fn allowed_for_user(
-        user_data: &UserData,
-        restrictions: &Restrictions,
-    ) -> AudioItemAvailability {
-        let country = &user_data.country;
-        let user_catalogue = match user_data.attributes.get("catalogue") {
-            Some(catalogue) => catalogue,
-            None => "premium",
-        };
-
-        for premium_restriction in restrictions.iter().filter(|restriction| {
-            restriction
-                .catalogue_strs
-                .iter()
-                .any(|restricted_catalogue| restricted_catalogue == user_catalogue)
-        }) {
-            if let Some(allowed_countries) = &premium_restriction.countries_allowed {
-                // A restriction will specify either a whitelast *or* a blacklist,
-                // but not both. So restrict availability if there is a whitelist
-                // and the country isn't on it.
-                if allowed_countries.iter().any(|allowed| country == allowed) {
-                    return Ok(());
-                } else {
-                    return Err(UnavailabilityReason::NotWhitelisted);
+                if track.duration <= 0 {
+                    return Err(Error::unavailable(MetadataError::InvalidDuration(
+                        track.duration,
+                    )));
                 }
-            }
 
-            if let Some(forbidden_countries) = &premium_restriction.countries_forbidden {
-                if forbidden_countries
+                if track.is_explicit {
+                    if let Some(value) = session.get_user_attribute("filter-explicit-content") {
+                        if &value == "1" {
+                            return Err(Error::unavailable(MetadataError::ExplicitContentFiltered));
+                        }
+                    }
+                }
```

Comment:
Maybe this could be turned into a method on the `Session`. Something like `session.filter_explicit_content()`. Then, this could be reused below. Alternatively, you could handle it the same as the `image-url` user attribute.

Reply:
No problem I can do that.